src/utils/mtev_lfu.c

```diff
@@ -234,22 +231,30 @@ touch_lfu_cache_no_lock(mtev_lfu_t *c, struct lfu_entry *e)
 {
   if (c->max_entries > 0) {
     struct lfu_cache_entry *bucket = e->frequency_list_head;
-    struct lfu_cache_entry *next_bucket = STAILQ_NEXT(bucket, list_entry);
+    struct lfu_cache_entry *next_bucket = TAILQ_NEXT(bucket, list_entry);
+    struct lfu_cache_entry *prev_bucket = TAILQ_PREV(bucket, frequency_list, list_entry);
     size_t bucket_freq = bucket->frequency;
     mtev_boolean need_new_bucket = mtev_false;
 
     if (next_bucket == NULL ||
-        next_bucket == STAILQ_FIRST(&c->lfu_frequency_list) ||
+        next_bucket == TAILQ_FIRST(&c->lfu_frequency_list) ||
         next_bucket->frequency != bucket_freq + 1) {
       need_new_bucket = mtev_true;
     }
     struct lfu_cache_entry *empty = remove_from_frequency_list_no_lock(c, e);
     if (need_new_bucket) {
       if (empty != NULL) {
+        /* reuse this bucket instead of allocating */
+        empty->frequency = bucket_freq + 1;
+        if (next_bucket && next_bucket->frequency < empty->frequency) {
```

Comment:
this case should be impossible? `c->lfu_frequency_list` is now stored sorted, so `empty`'s old frequency is guaranteed to be less than `next_bucket->frequency`. Further, we've already checked if `next_bucket->frequency` == new frequency (i.e. old frequency + 1), and it's not, because we `need_new_bucket`... I think we're OK w/ replacing all this logic with the body of the `else` block.